### PR TITLE
SQL: Introduce HISTOGRAM grouping function

### LINKED ISSUE #36509: SQL: Introduce Histogram function
A powerful, popular bucketing function in Elasticsearch is that of histograms. Such an ability should be available in SQL as well, allowing users to break down data by ranges effectively grouping by a range.

```sql
SELECT HISTOGRAM(number, 50) AS h FROM index GROUP BY h
SELECT HISTOGRAM(date, INTERVAL 1 YEAR) AS h FROM index GROUP BY h
```

### PR DESCRIPTION
**Pull request by costin**
Introduce Histogram grouping function for bucketing/grouping data based
 on a given range. Both date and numeric histograms are supported using
 the appropriate range declaration (numbers vs intervals).

```sql
SELECT HISTOGRAM(number, 50) AS h FROM index GROUP BY h
SELECT HISTOGRAM(date, INTERVAL 1 YEAR) AS h FROM index GROUP BY h
```

In addition add multiply operator for Intervals
Add docs for intervals and histogram

Fix #36509

Note: Histograms don't work as arguments to functions nor inside Having yet. That's planned as a separate PR.

### CODE REVIEW

**Review by astefan — COMMENTED** [review]
Left few cosmetic comments. Still review-in-progress.

File: x-pack/plugin/sql/qa/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/JdbcDocCsvSpecIT.java [hunk 1]
```diff
@@ -68,8 +68,8 @@ protected void assertResults(ResultSet expected, ResultSet elastic) throws SQLEx
         //
         // uncomment this to printout the result set and create new CSV tests
         //
-        //JdbcTestUtils.logLikeCLI(elastic, log);
-        JdbcAssert.assertResultSets(expected, elastic, log, true);
+        JdbcTestUtils.logLikeCLI(elastic, log);
```
Comment (matriv) [inline x-pack/plugin/sql/qa/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/JdbcDocCsvSpecIT.java]: revert this change?

File: x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/DebugSqlSpec.java [hunk 2]
```diff
@@ -16,7 +16,7 @@
     @ParametersFactory(shuffle = false, argumentFormatting = PARAM_FORMATTING)
     public static List<Object[]> readScriptSpec() throws Exception {
         Parser parser = specParser();
-        return readScriptSpec("/debug.sql-spec", parser);
+        return readScriptSpec("/datetime.sql-spec", parser);
```
Comment (matriv) [inline x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/DebugSqlSpec.java]: please revert

File: x-pack/plugin/sql/qa/src/main/resources/agg.csv-spec [hunk 3]
```diff
@@ -137,3 +137,80 @@ null           |null           |null           |null           |null           |
...
 ;
+
+
+//
+// Grouping functions
+//
+
+
+histogramNumeric
+SELECT HISTOGRAM(salary, 5000) AS h FROM test_emp GROUP BY h;
```
Comment (matriv) [inline x-pack/plugin/sql/qa/src/main/resources/agg.csv-spec]: don't we need ORDER BY to avoid flakiness?

File: x-pack/plugin/sql/qa/src/main/resources/agg.csv-spec [hunk 4]
```diff
@@ -137,3 +137,80 @@ null           |null           |null           |null           |null           |
...
+50000          
+55000          
+60000          
+65000          
+70000  
+;
+
+histogramDate
+schema::h:ts|c:l
+SELECT HISTOGRAM(birth_date, INTERVAL 1 YEAR) AS h, COUNT(*) as c FROM test_emp GROUP BY h;
```
Comment (matriv) [inline x-pack/plugin/sql/qa/src/main/resources/agg.csv-spec]: same here

File: x-pack/plugin/sql/qa/src/main/resources/agg.csv-spec [hunk 5]
```diff
@@ -137,3 +137,80 @@ null           |null           |null           |null           |null           |
...
+1956-03-15T00:00:00Z|4              
+1955-03-21T00:00:00Z|4              
+1954-03-26T00:00:00Z|7              
+1953-03-31T00:00:00Z|10             
+1952-04-05T00:00:00Z|10             
+1951-04-11T00:00:00Z|1              
+null                |10   
+;
+
+histogramDateWithDateFunction-Ignore
```
Comment (matriv) [inline x-pack/plugin/sql/qa/src/main/resources/agg.csv-spec]: This is the current restriction that the histogram cannot be used as a function arg?
Will it be supported in the future?

File: x-pack/plugin/sql/qa/src/main/resources/datetime-interval.csv-spec [hunk 6]
```diff
@@ -20,119 +20,177 @@ SELECT INTERVAL 1 YEARS AS y, INTERVAL 2 MONTHS AS m, INTERVAL 3 DAYS AS d, INTE
...
+minuteSecond
 SELECT INTERVAL '163:59.163' MINUTE TO SECOND AS interval;
 
 interval
 ---------------
 +0 02:43:59.163  
 ;
 
-testDatePlusInterval
+intervalPlusInterval
```
Comment (matriv) [inline x-pack/plugin/sql/qa/src/main/resources/datetime-interval.csv-spec]: Maybe add `Local` suffix to all of them since further below there are tests on table columns.

File: x-pack/plugin/sql/qa/src/main/resources/agg.csv-spec [hunk 7]
```diff
@@ -137,3 +137,80 @@ null           |null           |null           |null           |null           |
...
 4              |4              |72             |4              |4.0            |0.0            |NaN            |NaN   
 ;
+
+
+//
+// Grouping functions
+//
+
+
+histogramNumeric
```
Comment (matriv) [inline x-pack/plugin/sql/qa/src/main/resources/agg.csv-spec]: I think we can remove these 2 tests since they are also in the docs csv spec.

File: x-pack/plugin/sql/qa/src/main/resources/datetime-interval.csv-spec (hunk 6 above)
Comment (matriv) [reply x-pack/plugin/sql/qa/src/main/resources/datetime-interval.csv-spec]: Maybe we should remove them from here as they are also tested in the docs csv spec.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java [hunk 8]
```diff
@@ -108,7 +108,11 @@ public Analyzer(Configuration configuration, FunctionRegistry functionRegistry,
                 new ResolveAggsInHaving()
                 //new ImplicitCasting()
                 );
-        return Arrays.asList(substitution, resolution);
+        Batch finish = new Batch("Finish Analysis",
+                new PruneSubqueryAliases(),
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java]: Why don't we merge them in one rule?

File: x-pack/plugin/sql/qa/src/main/resources/math.csv-spec [hunk 9]
```diff
@@ -99,31 +99,31 @@ SELECT MIN(salary) mi, MAX(salary) ma, YEAR(hire_date) year, ROUND(AVG(languages
...
       mi:i     |      ma:i     |     year:i    |ROUND(AVG(languages),1):d|TRUNCATE(AVG(languages),1):d|   COUNT(1):l    
 ---------------+---------------+---------------+-------------------------+----------------------------+---------------
-25324          |70011          |1987           |3.0                      |3.0                         |15             
-25945          |73578          |1988           |2.9                      |2.8                         |9              
-25976          |74970          |1989           |3.0                      |3.0                         |13             
-31120          |71165          |1990           |3.1                      |3.0                         |12             
-30404          |58715          |1993           |3.0                      |3.0                         |3              
-35742          |67492          |1994           |2.8                      |2.7                         |4              
-45656          |45656          |1996           |3.0                      |3.0                         |1           
+25324          |70011          |1986           |3.0                      |3.0                         |15
```
Comment (matriv) [inline x-pack/plugin/sql/qa/src/main/resources/math.csv-spec]: where is this `-1 year` change come from?

File: docs/reference/sql/functions/date-time.asciidoc [hunk 10]
```diff
@@ -1,7 +1,99 @@
...
+[cols="^,^",options="header"]
+
+|===
+|  Interval                                     | Description
+                                              
+| `INTERVAL '1-2' YEAR TO MONTH`                | 1 year and 2 months
+| `INTERVAL '3 4' DAYS TO HOURS`                | 3 days and 4 hours
+| `INTERVAL '5 6:12' DAYS TO MINUTES`           | 5 days, 6 hours and 12 minutes
+| `INTERVAL '3 4:56:01' DAY TO SECOND`          | 3 days, 4 hours, 56 minutes and 1 second
+| `INTERVAL '2 3:45:01.23456789' DAY TO SECOND` | 2 days, 3 hours, 45 minutes, 1 second and 234567890 milliseconds
```
Comment (matriv) [inline docs/reference/sql/functions/date-time.asciidoc]: this should be nanoseconds, also below.

**Review by matriv — COMMENTED** [review]
Nice work! Left a first batch of comments.

**Review by matriv — COMMENTED** [review]
I'm through with the review, nice work overall!

1. Maybe we could add an integ test for interval multiplication on table columns.
2. What happens if the histogram is used in the WHERE clause? should we catch this case?

File: docs/reference/sql/concepts.asciidoc [hunk 11]
```diff
@@ -64,4 +64,4 @@ Multiple clusters, each with its own namespace, connected to each other in a fed
 
 |===
 
-As one can see while the mapping between the concepts are not exactly one to one and the semantics somewhat different, there are more things in common than differences. In fact, thanks to SQL declarative nature, many concepts can move across {es} transparently and the terminology of the two likely to be used interchangeably through-out the rest of the material.
+As one can see while the mapping between the concepts are not exactly one to one and the semantics somewhat different, there are more things in common than differences. In fact, thanks to SQL declarative nature, many concepts can move across {es} transparently and the terminology of the two likely to be used interchangeably through-out the rest of the material.
```
[the remark this answers is not in the document]
Comment (costin) [reply docs/reference/sql/concepts.asciidoc]: Might be some whitespace - I've fixed `through-out`.

File: docs/reference/sql/functions/date-time.asciidoc [hunk 12]
```diff
@@ -1,7 +1,99 @@
...
 [[sql-functions-datetime]]
-=== Date and Time Functions
+=== Date/Time and Interval Functions and Operators
+
+{es-sql} offers a wide range of facilities for performing date/time manipulations.
+
+[[sql-functions-datetime-interval]]
+==== Intervals
+
+A common requirement when dealing with date/time in general is date/time math which resolves around
```
[the remark this answers is not in the document]
Comment (costin) [reply docs/reference/sql/functions/date-time.asciidoc]: revolves

File: docs/reference/sql/functions/grouping.asciidoc [hunk 13]
```diff
@@ -0,0 +1,53 @@
+[role="xpack"]
+[testenv="basic"]
+[[sql-functions-grouping]]
+=== Grouping Functions
+
+Functions for creating special __grouping__s (also known as _bucketing_) and need to be used
```
[the remark this answers is not in the document]
Comment (costin) [reply docs/reference/sql/functions/grouping.asciidoc]: rephrased it.

File: docs/reference/sql/functions/grouping.asciidoc [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+`Histogram` can be applied on either numeric fields:
+
+
+["source","sql",subs="attributes,callouts,macros"]
+----
+include-tagged::{sql-specs}/docs.csv-spec[histogramNumeric]
+----
+
+and date/time fields:
```
[the remark this answers is not in the document]
Comment (costin) [reply docs/reference/sql/functions/grouping.asciidoc]: fixed

File: docs/reference/sql/functions/date-time.asciidoc [hunk 15]
```diff
@@ -1,7 +1,99 @@
 [role="xpack"]
 [testenv="basic"]
 [[sql-functions-datetime]]
-=== Date and Time Functions
+=== Date/Time and Interval Functions and Operators
+
```
[the remark this answers is not in the document]
Comment (costin) [reply docs/reference/sql/functions/date-time.asciidoc]: done.

File: docs/reference/sql/functions/date-time.asciidoc [hunk 16]
```diff
@@ -1,7 +1,99 @@
...
+{es-sql} offers a wide range of facilities for performing date/time manipulations.
+
+[[sql-functions-datetime-interval]]
+==== Intervals
+
+A common requirement when dealing with date/time in general is date/time math which resolves around 
+the notion of ``interval``s, a topic that is worth exploring in the context of {es} and {es-sql}.
+
+{es} has comprehensive support for <<date-math, date math>> both inside <<date-math-index-names,index names>> and <<mapping-date-format, queries>>.
```
[the remark this answers is not in the document]
Comment (costin) [reply docs/reference/sql/functions/date-time.asciidoc]: done.

File: docs/reference/sql/functions/grouping.asciidoc [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+HISTOGRAM ( date_exp<3>,  date_time_interval<4>)
+----
+
+*Input*:
+
+<1> numeric expression (typically a field)
+<2> numeric interval
+<3> date/time expression (typically a field)
+<4> date/time <<sql-functions-datetime-interval, interval>>
+...
```
[the remark this answers is not in the document]
Comment (costin) [reply docs/reference/sql/functions/grouping.asciidoc]: left-over.

File: docs/reference/sql/functions/grouping.asciidoc [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+
+`Histogram` can be applied on either numeric fields:
+
+
+["source","sql",subs="attributes,callouts,macros"]
+----
+include-tagged::{sql-specs}/docs.csv-spec[histogramNumeric]
+----
+
+and date/time fields:
```
[the remark this answers is not in the document]
Comment (costin) [reply docs/reference/sql/functions/grouping.asciidoc]: done.

File: x-pack/plugin/sql/qa/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/JdbcDocCsvSpecIT.java (hunk 1 above)
Comment (costin) [reply x-pack/plugin/sql/qa/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/JdbcDocCsvSpecIT.java]: Fixed.

File: x-pack/plugin/sql/qa/src/main/resources/agg.csv-spec (hunk 3 above)
Comment (costin) [reply x-pack/plugin/sql/qa/src/main/resources/agg.csv-spec]: No since all groups have an implicit order (due to how composite agg works).

File: x-pack/plugin/sql/qa/src/main/resources/agg.csv-spec (hunk 4 above)
Comment (costin) [reply x-pack/plugin/sql/qa/src/main/resources/agg.csv-spec]: See above.

File: x-pack/plugin/sql/qa/src/main/resources/agg.csv-spec (hunk 5 above)
Comment (costin) [reply x-pack/plugin/sql/qa/src/main/resources/agg.csv-spec]: I plan to look at it in the following days.

File: x-pack/plugin/sql/qa/src/main/resources/datetime-interval.csv-spec (hunk 6 above)
Comment (costin) [reply x-pack/plugin/sql/qa/src/main/resources/datetime-interval.csv-spec]: I'd like to keep things separate - while there is some duplication the doc lifecycle is different from that of the test. In fact, docs were added to avoid having obscure queries that make sense for testing show up in our documentation.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java (hunk 8 above)
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java]: Because they have different concerns - I would expect the second one to change when we add full support for subselects.

File: x-pack/plugin/sql/qa/src/main/resources/math.csv-spec (hunk 9 above)
Comment (costin) [reply x-pack/plugin/sql/qa/src/main/resources/math.csv-spec]: Relative vs absolute time computing. ES allows both absolute offsets (24h) or relative (1d) which takes into account if the date/time is within a day.
Based on the intervals given, ES keeps going back and forth but that's surprising and unexpected - in SQL as the intervals vary, I've used the absolute one so that `INTERVAL 1 YEAR` and `INTERVAL 12 MONTHS`, etc... gives you the same results.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/FunctionRegistry.java [hunk 19]
```diff
@@ -415,6 +418,28 @@ public boolean functionExists(String functionName) {
         T build(Location location, Expression target, TimeZone tz);
     }
 
+    /**
+     * Build a {@linkplain FunctionDefinition} for a plur/n-ary function that
```
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/FunctionRegistry.java]: Initially I wanted to go for multi/variable arguments but that's not the case. Will update it.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/grouping/GroupingFunction.java [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+
+    @Override
+    protected Pipe makePipe() {
+        // unresolved AggNameInput (should always get replaced by the folder)
+        return new AggNameInput(location(), this, name());
+    }
+
+    @Override
+    public ScriptTemplate asScript() {
+        throw new SqlIllegalArgumentException("Grouping functions cannot be scripted");
```
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/grouping/GroupingFunction.java]: That implies an operation that somehow was accepted but it's not implemented, which I would consider a bug.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/RestSqlClearCursorAction.java [hunk 21]
```diff
@@ -27,7 +27,7 @@
 
     private static final DeprecationLogger deprecationLogger = new DeprecationLogger(LogManager.getLogger(RestSqlClearCursorAction.class));
 
-    RestSqlClearCursorAction(Settings settings, RestController controller) {
+    public RestSqlClearCursorAction(Settings settings, RestController controller) {
```
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/RestSqlClearCursorAction.java]: No, this change shouldn't have occurred in the first place but the PR got it before I got a chance to review it.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/operator/arithmetic/BinaryArithmeticProcessor.java [hunk 22]
```diff
@@ -82,7 +82,28 @@ default Object wrap(Object l, Object r) {
             throw new SqlIllegalArgumentException("Cannot compute [-] between [{}] [{}]", l.getClass().getSimpleName(),
                     r.getClass().getSimpleName());
         }, "-"),
-        MUL(Arithmetics::mul, "*"),
+        MUL((Object l, Object r) -> {
+            if (l instanceof Number && r instanceof Number) {
+                return Arithmetics.mul((Number) l, (Number) r);
+            }
+            l = unwrapJodaTime(l);
```
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/operator/arithmetic/BinaryArithmeticProcessor.java]: See `BinaryArithmeticTests`

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/grouping/Histogram.java [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+
+    @Override
+    public DataType dataType() {
+        return field().dataType();
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/grouping/Histogram.java]: Why the `replaceChildren(List<Expression>)` here and not defined as `final` in `GroupingFunction` enforcing the replacement of a single children? `GroupingFunction` would then define a `protected abstract GroupingFunction replaceChildren(Expression newExpression);`. Are we expecting more grouping functions with variable list of children in the future?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java [hunk 24]
```diff
@@ -537,10 +537,31 @@ public Literal visitIntervalLiteral(IntervalLiteralContext ctx) {
 
         DataType intervalType = Intervals.intervalType(source(interval), leading, trailing);
 
-        boolean negative = interval.sign != null && interval.sign.getType() == SqlBaseParser.MINUS;
+        // negation outside the interval - use xor
+        boolean negative = false;
+
+        ParserRuleContext parentCtx = interval.getParent();
+        if (parentCtx != null) {
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java]: What is this nested `if` block doing?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/agg/GroupByNumericHistogram.java [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+    @Override
+    public int hashCode() {
+        return Objects.hash(super.hashCode(), interval);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (super.equals(obj)) {
+            GroupByNumericHistogram other = (GroupByNumericHistogram) obj;
+            return Objects.equals(interval, other.interval);
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/agg/GroupByNumericHistogram.java]: Why `Objects.equals` and not `interval == other.interval`?

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java [hunk 26]
```diff
@@ -392,4 +397,34 @@ public void testTranslateMathFunction_HavingClause_Painless() {
         assertThat(aggFilter.scriptTemplate().params().toString(), startsWith("[{a=MAX(int){a->"));
         assertThat(aggFilter.scriptTemplate().params().toString(), endsWith(", {v=10}]"));
     }
-}
+
+    public void testGroupByDateHistogram() {
+        LogicalPlan p = plan("SELECT MAX(int) FROM test GROUP BY HISTOGRAM(int, 1000)");
```
Comment (astefan) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java]: Do we have a situation where the histogram is being grouped by very small intervals? Like 1 second for a date interval.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/FunctionRegistry.java [hunk 27]
```diff
@@ -415,6 +418,28 @@ public boolean functionExists(String functionName) {
...
+     * operates on a datetime.
+     */
+    @SuppressWarnings("overloads") // These are ambiguous if you aren't using ctor references but we always do
+    static <T extends Function> FunctionDefinition def(Class<T> function, DatetimeBinaryFunctionBuilder<T> ctorRef, String... names) {
```
[the remark this answers is not in the document]
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/FunctionRegistry.java]: hm, thx! I missed something there, sorry.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java (hunk 24 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java]: Because of the structure of the grammar it's a way to get the `-` symbol from the immediate parent context in the hierarchy that could possible contain it.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/grouping/GroupingFunction.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/grouping/GroupingFunction.java]: But this is somehow the case here, no? We end up calling this method while we shouldn't. The `SqlIllegalArgumentException`  marks calling something (e.g.: sql function from the user perspective, or internal method with wrong artuments). Anyway, this is out of this PR's scope.

File: x-pack/plugin/sql/qa/src/main/resources/agg.csv-spec [hunk 28]
```diff
@@ -137,3 +137,80 @@ null           |null           |null           |null           |null           |
...
+1961-02-17T00:00:00Z|8
+1962-02-12T00:00:00Z|6
+1963-02-07T00:00:00Z|7
+1964-02-02T00:00:00Z|5
+         
+;
+
+histogramDateWithCountAndOrder
+schema::h:ts|c:l
+SELECT HISTOGRAM(birth_date, INTERVAL 1 YEAR) AS h, COUNT(*) as c FROM test_emp GROUP BY h ORDER BY h DESC;
```
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/sql/qa/src/main/resources/agg.csv-spec]: No bucket is returned (as that would break the notion of null group in SQL).

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/grouping/GroupingFunctionAttribute.java [hunk 29]
```diff
@@ -0,0 +1,74 @@
...
+import org.elasticsearch.xpack.sql.expression.function.FunctionAttribute;
+import org.elasticsearch.xpack.sql.tree.Location;
+import org.elasticsearch.xpack.sql.tree.NodeInfo;
+import org.elasticsearch.xpack.sql.type.DataType;
+
+import java.util.Objects;
+
+public class GroupingFunctionAttribute extends FunctionAttribute {
+
+    private final String propertyPath;
```
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/grouping/GroupingFunctionAttribute.java]: I've removed it

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/grouping/Histogram.java (hunk 23 above)
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/grouping/Histogram.java]: We don't know what the future will bring - the pattern above is used for `Unary[..]Function`s which a grouping function might be, but we don't know yet.
Adding it now results in 2 methods without any reuse so it's arguable whether it makes sense to do it or not. In general, I prefer to add a pattern when it applies to at least 2 instances, otherwise it's more of a premature recognition.
I've added it nevertheless.

File: x-pack/plugin/sql/qa/src/main/resources/datetime-interval.csv-spec [hunk 30]
```diff
@@ -20,119 +20,177 @@ SELECT INTERVAL 1 YEARS AS y, INTERVAL 2 MONTHS AS m, INTERVAL 3 DAYS AS d, INTE
...
+SELECT INTERVAL 1 DAY + INTERVAL 53 MINUTES;
+
+INTERVAL 1 DAY + INTERVAL 53 MINUTES
+------------------------------------
++1 00:53:00.0    
+;
+
+datePlusIntervalInline
```
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/sql/qa/src/main/resources/datetime-interval.csv-spec]: In this csv, `constant` is not used since for intervals that means repeating for every test so it becomes noise. Considering it has its own namespace (dedicated file) I think it's fine as is.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java (hunk 26 above)
Comment (costin) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java]: I tried a test with 1 second and it worked :). Our data might not be representative but I think it's a side effect of composite aggs which paginates through data instead of creating all the buckets up-front.

File: x-pack/plugin/sql/qa/src/main/resources/agg.csv-spec (hunk 28 above)
[the remark this answers is not in the document]
Comment (astefan) [reply x-pack/plugin/sql/qa/src/main/resources/agg.csv-spec]: Can you mention this in the documentation, please? Users used to `date_histogram` (where the "missing" buckets are still returned by ES) will probably expect the same from HISTOGRAM.

**Review by matriv — APPROVED** [review]
LGTM! Great work!

**Review by astefan — APPROVED** [review]
Great stuff! LGTM
